app/supplementary_data.py

```diff
@@ -0,0 +1,89 @@
+import json
+from typing import Mapping
+
+from flask import current_app
+from marshmallow import ValidationError
+from requests import RequestException
+from structlog import get_logger
+
+from app.utilities.request_session import get_retryable_session
+from app.utilities.supplementary_data_parser import validate_supplementary_data_v1
+
+SUPPLEMENTARY_DATA_URL = "http://localhost:5003/v1/unit_data"
+SUPPLEMENTARY_DATA_REQUEST_BACKOFF_FACTOR = 0.2
+SUPPLEMENTARY_DATA_REQUEST_MAX_RETRIES = 2  # Totals no. of request should be 3. The initial request + SUPPLEMENTARY_DATA_REQUEST_MAX_RETRIES
+SUPPLEMENTARY_DATA_REQUEST_TIMEOUT = 3
+SUPPLEMENTARY_DATA_REQUEST_RETRY_STATUS_CODES = [
+    408,
+    429,
+    500,
+    502,
+    503,
+    504,
+]
+
+logger = get_logger()
+
+
+class SupplementaryDataRequestFailed(Exception):
+    def __str__(self) -> str:
+        return "Supplementary Data request failed"
+
+
+def get_supplementary_data(dataset_id: str, unit_id: str, survey_id: str) -> dict:
+    supplementary_data_url = current_app.config["SDS_API_BASE_URL"]
+
+    constructed_supplementary_data_url = (
+        f"{supplementary_data_url}?dataset_id={dataset_id}&unit_id={unit_id}"
+    )
+
+    session = get_retryable_session(
+        max_retries=SUPPLEMENTARY_DATA_REQUEST_MAX_RETRIES,
+        retry_status_codes=SUPPLEMENTARY_DATA_REQUEST_RETRY_STATUS_CODES,
+        backoff_factor=SUPPLEMENTARY_DATA_REQUEST_BACKOFF_FACTOR,
+    )
+
+    try:
+        response = session.get(
+            constructed_supplementary_data_url,
+            timeout=SUPPLEMENTARY_DATA_REQUEST_TIMEOUT,
+        )
+    except RequestException as exc:
+        logger.exception(
+            "Error requesting supplementary data",
+            supplementary_data_url=constructed_supplementary_data_url,
+        )
+        raise SupplementaryDataRequestFailed from exc
+
+    if response.status_code == 200:
+        supplementary_data_response_content = response.content.decode()
+        supplementary_data = json.loads(supplementary_data_response_content)
+
+        return validate_supplementary_data(
+            supplementary_data=supplementary_data,
+            dataset_id=dataset_id,
+            unit_id=unit_id,
+            survey_id=survey_id,
+        )
+
+    logger.error(
+        "got a non-200 response for supplementary data request",
+        status_code=response.status_code,
+        schema_url=constructed_supplementary_data_url,
+    )
+
+    raise SupplementaryDataRequestFailed
```

Comment:
Add a runner error handler for this to return our 500 error page if this happens. Needs testing.

Reply:
Updated and added a test 👍